### ISSUE TITLE
how to set random seed for BucketIterator to guarantee that it produce the same itrator every time you run the code ?

### ISSUE THREAD

**Issue by zide05** [pytorch/text#522]

**Comment by PetrochukM** [comment]
You mean something like this? https://github.com/PetrochukM/PyTorch-NLP#deterministic-functions

**Comment by ajason08** [comment]
Yes, like  https://github.com/PetrochukM/PyTorch-NLP#deterministic-functions
but PyTorch built-in.
does it exist?

**Comment by zhangguanheng66** [comment]
I think you have to use `DataLoader`, which is only compatible with those datasets under the new dataset abstraction. Which datasets are you using now? @ajason08

**Comment by ajason08** [comment]
At this moment I am using the Dataset and TabularDataset to load my data.
However, my main concern, which brought me here was trying to fully reproduce my results. I set all the seeds (code at the end), still I am getting 3% variation in F1 at the end of  the experiment. This is a simple single-layer CNN with no additional libraries. (I can upload the code too).

I think that maybe the BucketIterator is the source of randomness, due this snippet is returning different results.  I also tried shuffle=False and using Iterator instead of BucketIterator.
```
train_iterator, valid_iterator, dev_iterator = data.BucketIterator.splits(
  (train_data, valid_data, dev_data), 
  sort = False, 
  batch_size = BATCH_SIZE, 
  device = device
  #shuffle=False
  )
for batch in train_iterator:      
  x = model(batch.tweet)
  print (x) # I am receiving different results each time I run this snippet
  break
```

**Setting seeds:**
```
def seed_torch(seed=1234):
    os.environ['PYTHONHASHSEED'] = str(seed)
    random.seed(seed)    
    np.random.seed(seed)
    np.random.RandomState(seed)

    torch.manual_seed(seed) 
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed) #seed all gpus    
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.enabled = False  
    torch.backends.cudnn.benchmark = False
```
What do you think should I do in order to get the same results from BucketIterator, and hopefully reproduce my results? Thank you!

**Comment by zhangguanheng66** [comment]
@mttk @bentrevett Any ideas with `BucketIterator`?

**Comment by bencwallace** [comment]
It seems that, currently, you can't set the random seed for a BucketIterator. It looks like [this](https://github.com/pytorch/text/blob/master/torchtext/data/iterator.py#L76) line would have to be changed to accept a random seed. I could give it a shot. What's the policy on submitting PRs for torchtext? Is it the same as for PyTorch?

**Comment by zhangguanheng66** [comment]
> It seems that, currently, you can't set the random seed for a BucketIterator. It looks like [this](https://github.com/pytorch/text/blob/master/torchtext/data/iterator.py#L76) line would have to be changed to accept a random seed. I could give it a shot. What's the policy on submitting PRs for torchtext? Is it the same as for PyTorch?

Just FYI, The `Iterator` will be retired after October as we switch to `torch.utils.data`.

**Comment by bencwallace** [comment]
Good to know! Does that mean `BucketIterator` and `BPTTIterator` will inherit directly from PyTorch's `Dataset`? They're both very useful iterators!

**Comment by zhangguanheng66** [comment]
We will switch to the utils in `torch.utils.data`, for example `DataLoader`. See some discussions https://github.com/pytorch/text/issues/664

**Comment by bencwallace** [comment]
Thanks for the reference. In case anyone else comes across this issue in the meantime, I realized you actually *can* control the random shuffle by calling `random.setstate` (with a desired state) before calling `BucketIterator.splits`.

**Comment by abhinavarora** [comment]
Closing issue as we have gotten rid of iterators and the issue is no more relevant. 
#1537
